### PR TITLE
Offer o2::dataformats::TimeStamp class globally

### PR DESCRIPTION
**Pull request by sawenzel**
Moving a TimeStamp class from TPC to SimulationDataFormat
to be reused by other detectors;
This class (or its subclass TimeStampWithError) can be used
in places where previously FairTimeStamp was used.

The TimeStamp class is configurable on the type.

With this move, a couple of cleanup changes where possible:

  * use new class for tof::Digit
  * use new class for tpc::Cluster
  * remove unused FairTimeStamp header includes and
    fix some resulting missing includes that came
    indirectly with FairTimeStamp

### CODE REVIEW

File: Detectors/TOF/base/include/TOFBase/Digit.h [hunk 1]
```diff
@@ -23,12 +23,12 @@ namespace o2 {
 namespace tof {
 /// \class Digit
 /// \brief TOF digit implementation
-class Digit : public FairTimeStamp {
+class Digit : public o2::dataformats::TimeStamp<double> {
```
Comment (sawenzel) [inline Detectors/TOF/base/include/TOFBase/Digit.h]: @noferini : Can you check if ok with you?

File: Detectors/TPC/reconstruction/include/TPCReconstruction/Cluster.h [hunk 2]
```diff
@@ -16,27 +16,16 @@
...
-  float getTimeStampError() const {return mTimeStampError;}
-  void setTimeStamp(float t) {mTimeStamp = t;}
-  void setTimeStampError(float te) {mTimeStampError = te;}
- private:
-  float mTimeStamp = 0.f;
-  float mTimeStampError = 0.f;
-  ClassDefNV(ClusterTimeStamp, 1);
-};
-  
+using ClusterTimeStamp = o2::dataformats::TimeStampWithError<float>;
```
Comment (sawenzel) [inline Detectors/TPC/reconstruction/include/TPCReconstruction/Cluster.h]: @davidrohr : Is this ok?

File: DataFormats/simulation/include/SimulationDataFormat/TimeStamp.h [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  TimeStamp(T time) { mTimeStamp = time; }
+  T getTimeStamp() const { return mTimeStamp; }
+  void setTimeStamp(T t) { mTimeStamp = t; }
+ private:
+  T mTimeStamp = 0;
+  ClassDefNV(TimeStamp, 1);
+};
+
+template <typename T>
+class TimeStampWithError : public TimeStamp<T>
```
Comment (ktf) [inline DataFormats/simulation/include/SimulationDataFormat/TimeStamp.h]: Doesn't it make sense to have different precision for error and timestamp itself? Also for the sake of unit testing, why not adding a simple one which creates a `TimestampWithError` and retrieves its contents?

Comment (sawenzel) [reply DataFormats/simulation/include/SimulationDataFormat/TimeStamp.h]: If this is ever required we can add it later.

Comment (sawenzel) [reply DataFormats/simulation/include/SimulationDataFormat/TimeStamp.h]: Implemented suggestions.

**Comment by shahor02** [comment]
@sawenzel , I run my standard simulation with this class and got plenty of messages like
````
Warning in <TStreamerInfo::Build>: o2::TPC::Digit: base class o2::dataformats::TimeStamp<int> has no streamer or dictionary it will not be saved
````
and all the output branches with TimeStamp are empty. Strangely, this does not happen of for the error part of TimeStampWithError: the error info is stored but the not the time:
````
root [6] o2sim->Scan("TPCClusterHW.mTimeStamp : TPCClusterHW.mTimeStampError")
***********************************************
*    Row   * Instance * TPCCluste *  TPCClust *
***********************************************
*        0 *        0 *           *           *
*        1 *        0 *           * 0.9234804 *
*        1 *        1 *           * 0.8972870 *
*        1 *        2 *           * 0.8850445 *
*        1 *        3 *           * 0.8089578 *
*        1 *        4 *           * 0.2327176 *
*        1 *        5 *           * 0.8780974 *
*        1 *        6 *           * 0.9317334 *
*        1 *        7 *           * 0.8446878 *
````

Given that the dictionary generation in the SimulationDataLinkDef is done in a same way for both classes, I don't understand why they behave as they do. Moreover, that in the compilation I get:
````
[ 66%] Generating G__SimulationDataFormatDict.cxx, G__SimulationDataFormatDict_rdict.pcm, ../../lib/libSimulationDataFormat.rootmap
Warning: Unused class rule: o2::dataformats::TimeStampWithError<int, int>
Warning: Unused class rule: o2::dataformats::TimeStampWithError<double, double>
Warning: Unused class rule: o2::dataformats::TimeStampWithError<float, float>
Warning: Unused class rule: o2::dataformats::TimeStamp<int>
Warning: Unused class rule: o2::dataformats::TimeStamp<double>
Warning: Unused class rule: o2::dataformats::TimeStamp<float>
Scanning dependencies of target SimulationDataFormat
[ 66%] Building CXX object DataFormats/simulation/CMakeFiles/SimulationDataFormat.dir/G__SimulationDataFormatDict.cxx.o
````

Do you observe such thing?

BTW (sorry for late ignition), if TPC uses this class also in the reconstruction, would not it be better to move it to the new
directory  DataFormats/Detectors/Common, where I've already moved some stuff in https://github.com/AliceO2Group/AliceO2/pull/808 ?

Cheers,
 Ruben

**Comment by sawenzel** [comment]
The issue should be solved via #819. Thanks for spotting.

**Pull request merged**